js/DropZone/DropZone.js

```diff
@@ -0,0 +1,289 @@
+        this.options = options;
+        this.id = this.options.dropZoneId;
+        // cache methods with context
+        this.handleDropWithContext = this.handleDrop.bind(this);
+        this.handleWindowEnterWithContext = this.handleWindowEnter.bind(this);
+        this.handleWindowLeaveWithContext = this.handleWindowLeave.bind(this);
+        // a flag for determining support
+        this.supportsDataTransfer = true;
+        // setup files and size storage
+        this.files = [];
+        this.size = 0;
+    }
+
+    /**
+     * Initialise plugin
+     */
+    init () {
+        // add events for environments that support data transfer items
+        if (this.options.supported) {
+            this.windowActive = false;
+            this.dropZoneActive = false;
+            // prevent
+            this.eventManager.add(window, 'drag', this.eventManager.preventer);
+            this.eventManager.add(window, 'dragstart', this.eventManager.preventer);
+            this.eventManager.add(window, 'dragend', this.eventManager.preventer);
+            this.eventManager.add(window, 'dragover', this.eventManager.preventer);
+            this.eventManager.add(window, 'dragenter', this.eventManager.preventer);
+            this.eventManager.add(window, 'dragleave', this.eventManager.preventer);
+            this.eventManager.add(window, 'drop', this.eventManager.preventer);
+            // interactions
+            this.eventManager.add(window, 'dragenter', this.handleWindowEnterWithContext);
+            this.eventManager.add(window, 'dragleave', this.handleWindowLeaveWithContext);
+            this.eventManager.add(window, 'drop', this.handleDropWithContext);
+
+            // attempt to handle missed callbacks
+            // mouseout has proven to be more reliable than dragleave
+            this.eventManager.add(document, 'mouseout', this.idleTimer.start.bind(this, this.forceWindowLeave.bind(this)));
+        }
+    }
+
+    /**
+     * Handle drop event
+     * @param  {Event} event
+     */
+    handleDrop (event) {
+        const files = event.dataTransfer.items || event.dataTransfer.files;
+
+        // determine where the drop has taken place
+        // dropped on the DropZone
+        if (this.windowActive && this.dropZoneActive) {
+            this.addFiles(files);
+        // dropped on the window
+        } else if (this.windowActive && !this.dropZoneActive) {
+            this.callbackManager.create(
+                this.options.windowDrop,
+                this,
+                { files: this.files }
+            );
+        }
+
+        // reset DropZone state
+        this.windowActive = false;
+        this.dropZoneActive = false;
+    }
+
+    /**
+     * Validate and add files to the DropZone
+     * @param {Array} files
+     * @param {Object} meta
+     */
+    addFiles (files, meta = {}) {
+        const { valid, text } = this.validator.validate(files, this.files.length, this.size);
+        let processedFiles = [];
```

Comment:
Yep, not sure how that slipped through, probably didn't start life as an array 😄